### PR TITLE
show zero-width chars as '\uXXXX'

### PR DESCRIPTION
**Pull request by stevengj**
This PR changes `show` for `AbstractChar` to display printable but zero-width characters (typically combining characters) as `'\uXXXX'` rather than printing them, similar to non-printable characters.   Rationale: printing it is usually unreadable (e.g. a combining character gets combined with the leading `'`).

Before this PR, [U+0307](https://www.fileformat.info/info/unicode/char/0307/index.htm) “combining dot above” gets displayed as:
```jl
julia> '\u307'
'̇': Unicode U+0307 (category Mn: Mark, nonspacing)

julia> show(stdout, '\u307')
'̇'
```
where you can barely see the dot above the single quote.  After this PR, you get:
```jl
julia> '\u307'
'\u307': Unicode U+0307 (category Mn: Mark, nonspacing)

julia> show(stdout, '\u307')
'\u307'
```

### CODE REVIEW

File: NEWS.md [hunk 1]
```diff
@@ -87,6 +87,7 @@ New library features
 * Passing an IOBuffer as a stdout argument for Process spawn now works as
   expected, synchronized with `wait` or `success`, so a `Base.BufferStream` is
   no longer required there for correctness to avoid data-races ([#TBD]).
+* `show` for `AbstractChar` now displays zero-width characters as `'\uXXXX'` ([#52653]).
```
Comment (stevengj) [inline NEWS.md:90]:
```suggestion
* `show` for `AbstractChar` now displays zero-width characters as `'\uXXXX'` ([#52654]).
```

**Comment by stevengj** [comment]
(Another alternative would be to show it as `' ̇'`, i.e. to put the combining mark above a space.  However, this would necessitate changing the parser to accept literal characters in this format.)

It would be nice of if the long-format `text/plain` output (used by `display`) would show an example of the mark with some other character, e.g. something like
```
julia> '\u307'
'\u307': Unicode U+0307 (category Mn: Mark, nonspacing), e.g. □̇
```
but I'm not sure if it will be clear that `□` (or whatever character one chooses) is a placeholder.  Morever, it may not work correctly with all fonts/terminals, nor will it work with all combining characters (e.g. emoji modifiers only work after certain emoji).

File: test/char.jl [hunk 2]
```diff
@@ -189,6 +189,7 @@ end
     @test sprint(show, "text/plain", '$') == "'\$': ASCII/Unicode U+0024 (category Sc: Symbol, currency)"
     @test sprint(show, "text/plain", '$', context=:compact => true) == "'\$'"
     @test repr('$') == "'\$'"
+    @test repr('\u0307') == "'\u0307'" # zero-width combining char
```
Comment (stevengj) [inline test/char.jl:192]:
```suggestion
    @test repr('\u0307') == "'\\u307'" # zero-width combining char
```

**Comment by ericphanson** [comment]
For me, the PR text renders as
B9687199-61CF-4AAD-85DE-6F6B4B01BA7E
with the dot above the space, but it sounds like that isn’t the case in general. Do you mind including a screenshot?

**Comment by stevengj** [comment]
Marking as triage since it would be good to get a decision.

**Comment by LilithHafner** [comment]
Jeff wants to see the character if at all possible. I concur.

https://github.com/JuliaLang/julia/pull/52654#issuecomment-1872276348 looks nice, and https://github.com/JuliaLang/julia/pull/52654#issuecomment-1872522920 looks bad, but if there's not a reliable way of forcing the display to look like that then maybe it's not worth it. Scott thinks it's possible to do that very well.

`display` does a nice job.

There is a consensus that unicode is hard.

Overall, Jeff and Scott from triage think we shouldn't do this (Scott with a "for now" asterisk). Others abstained.

**Comment by stevengj** [comment]
> there's not a reliable way of forcing the display to look like that then maybe it's not worth it

A reliable way would be to show the combining character after a non-breaking space.  But that would require a corresponding change in the character-literal syntax.

**Pull request closed without merging**